sim/druid/druid.go

```diff
@@ -3,30 +3,201 @@ package druid
 import (
 	"time"
 
-	"github.com/wowsims/tbc/sim/core/proto"
 	"github.com/wowsims/tbc/sim/core"
+	"github.com/wowsims/tbc/sim/core/proto"
+	"github.com/wowsims/tbc/sim/core/stats"
 )
 
 type Druid struct {
 	core.Character
+	SelfBuffs
+	Talents proto.DruidTalents
+
+	innervateCD  time.Duration
+	naturesGrace bool // when true next spellcast is 0.5s faster
+
+	// cached cast stuff
+	starfireSpell         core.SingleTargetDirectDamageSpell
+	starfire8CastTemplate core.SingleTargetDirectDamageSpellTemplate
+	starfire6CastTemplate core.SingleTargetDirectDamageSpellTemplate
+
+	MoonfireSpell        core.DamageOverTimeSpell
+	moonfireCastTemplate core.DamageOverTimeSpellTemplate
+
+	wrathSpell        core.SingleTargetDirectDamageSpell
+	wrathCastTemplate core.SingleTargetDirectDamageSpellTemplate
+
+	InsectSwarmSpell        core.DamageOverTimeSpell
+	insectSwarmCastTemplate core.DamageOverTimeSpellTemplate
+}
+
+type SelfBuffs struct {
+	Omen      bool
+	Innervate bool
 }
 
 func (druid *Druid) GetCharacter() *core.Character {
 	return &druid.Character
 }
 
 func (druid *Druid) AddRaidBuffs(raidBuffs *proto.RaidBuffs) {
-	// TODO: Use talents to check for imp gotw
 	raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectRegular
+	if druid.Talents.ImprovedMarkOfTheWild > 0 { // ya ya whatever
+		raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectImproved
+	}
 }
+
+const ravenGoddessItemID = 32387
+
 func (druid *Druid) AddPartyBuffs(partyBuffs *proto.PartyBuffs) {
-	//buffs.Moonkin = proto.TristateEffect_TristateEffectRegular
-	// check for idol of raven goddess equipped
+	if druid.Talents.MoonkinForm { // assume if you have moonkin talent you are using it.
+		partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectRegular
+		for _, e := range druid.Equip {
+			if e.ID == ravenGoddessItemID {
+				partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectImproved
+				break
+			}
+		}
+	}
+}
+
+func (druid *Druid) Init(sim *core.Simulation) {
+	druid.starfire8CastTemplate = druid.newStarfireTemplate(sim, 8)
+	druid.starfire6CastTemplate = druid.newStarfireTemplate(sim, 6)
+	druid.moonfireCastTemplate = druid.newMoonfireTemplate(sim)
+	druid.wrathCastTemplate = druid.newWrathTemplate(sim)
+	druid.insectSwarmCastTemplate = druid.newInsectSwarmTemplate(sim)
 }
 
 func (druid *Druid) Reset(newsim *core.Simulation) {
+	druid.moonfireCastTemplate.Apply(&druid.MoonfireSpell)
+	druid.Character.Reset(newsim)
 }
 
+func (druid *Druid) Advance(sim *core.Simulation, elapsedTime time.Duration) {
+	druid.Character.RegenManaMP5Only(sim, elapsedTime)
```

Comment:
currently the regen from the resto tree means they never need to wait more than  1-2s before casting again.  I can use it but I found it wasn't needed

Reply:
Even while casting we want to include the 30% regen thats allowed to continue while casting. If I understand this setup correctly, you are indeed taking that into account inside character.Advance(). But it would be a lot more clear to keep all regen logic out of character.Advance() and use a different call here, like character.RegenMP5AndPartialSpiritOnly().

Reply:
that seems reasonable, I can rename the function